### ISSUE TITLE
jQuery not working in Gutenberg Editor view

### ISSUE THREAD

**Issue by munirkamal** [AdvancedCustomFields/acf#172]
Hi,

Have been playing around with ACFBlock functionality to create custom blocks. First of all hats of to the team ACF to make the block creation process soo easier.

Now I am building a simple Tabs block which uses jQuery to work. The block output on frontend works perfect but Tabs doesn't function/work in the editor. I am providing the code below, any help will be much appreciated. 

## Register Function

    add_action('acf/init', 'gutenberg_acf_blocks');
    function gutenberg_acf_blocks() {

       acf_register_block(array(
           'name'              => 'tabs',
           'mode'                => 'preview',
           'title'             => __('Tabs'),
           'description'       => __('A custom tabs block.'),
           'render_template'   => plugin_dir_path( __FILE__ ) . 'block-templates/tabs.php',
           'category'          => 'formatting',
           'enqueue_assets' => function(){
              wp_enqueue_style( 'acf-blocks-css', plugin_dir_url( __FILE__ ) . 'css/acfblocks.css' );
              wp_enqueue_script( 'acf-blocks-js', plugin_dir_url( __FILE__ ) . 'js/acfblocks.js', array('jquery'), '', false );
            },
       ));
    }
    }

## Render Code from render_template
    
<?php if( have_rows('tabs') :disappointed: ?>

    <div class="tabs">

    <ul class="tabs-list">
    <?php while( have_rows('tabs') :disappointed: the_row(); ?>
      <?php $tab_title = get_sub_field('tab_title'); ?>
      <li><a href="#tab-<?php echo get_row_index();?>"><?php echo $tab_title; ?></a></li>
    <?php endwhile; ?>
    </ul>

         
    <?php while( have_rows('tabs') :disappointed: the_row(); ?>

        <?php $tab_content = get_sub_field('tab_content'); ?>
      
          <div id="tab-<?php echo get_row_index();?>" class="tab">
               <?php echo $tab_content; ?>
          </div>

    <?php endwhile; ?>
    </div>

    <?php endif; ?>

## Custom JS code

    jQuery(document).ready(function($){

    $(".tabs-list li:first-child").addClass("active");
    $(".tabs .tab:first").addClass("active");

     $(".tabs-list li a").click(function(e){
         e.preventDefault();
     });

    $(".tabs-list li").click(function(){
     var tabid = $(this).find("a").attr("href");     
     $(".tabs-list li,.tabs div.tab").removeClass("active");
     $(".tab").hide();   // hiding open tab
     $(tabid).show();    // show tab
     $(this).addClass("active"); //  adding active class to clicked tab

    });

    });

Thanks

**Comment by elliotcondon** [comment]
Hi @munirkamal 

Adding JS to the block editor (back-end) requires some extra code for it to work. Please checkout the "Adding block scripts" example here: https://www.advancedcustomfields.com/resources/acf_register_block/#examples

**Comment by munirkamal** [comment]
Yeah, I have seen that example, by any chance do you have more examples? Or some 3rd party block you can refer me to?

**Comment by elliotcondon** [comment]
Hi @munirkamal 

This is currently the only code example for adding Javascript to an ACF Block.
If you are finding it difficult to understand, or having troubles implementing this technique, please let me know the details.

**Comment by munirkamal** [comment]
> Hi @munirkamal
> 
> This is currently the only code example for adding Javascript to an ACF Block.
> If you are finding it difficult to understand, or having troubles implementing this technique, please let me know the details.

@elliotcondon 
Well yes a little more help would be very much appreciated as I am kind of stuck making JS code run within Gutenberg editor for now. My Whole block code has been provided in this issue. Here is the JS part again. If you can help me make this code work inside Gutenberg editor that would be amazing. And this can also serve as an additional example for ACF users too.

```
 jQuery(document).ready(function($){

$(".tabs-list li:first-child").addClass("active");
$(".tabs .tab:first").addClass("active");

 $(".tabs-list li a").click(function(e){
     e.preventDefault();
 });


$(".tabs-list li").click(function(){
 var tabid = $(this).find("a").attr("href");     
 $(".tabs-list li,.tabs div.tab").removeClass("active");
 $(".tab").hide();   // hiding open tab
 $(tabid).show();    // show tab
 $(this).addClass("active"); //  adding active class to clicked tab

});

});
```
TIA

**Comment by elliotcondon** [comment]
Hi @munirkamal 

To get custom JS working in an ACF Block, a few requirements are needed:

- [ ] A callback function to initialize a single block hooked into the 'render_block_preview' action.
- [ ] Another clalback function to initialize all blocks hooked into the $(document).ready() function.

Your code ticks the second requirement, and would work on a"normal" website where are all the DOM elements exist on page load.

What you are missing is that first requirement, which is explained in detail (and code example) over on the documentation linked earlier.

I understand that you would prefer me to write the code for you, but this not going to happen. Instead, I will happily walk you through the steps so that you can tech yourself :)

Please read over the docs once more, and try to tick those boxes above, and let me know any specific questions you have.

**Comment by andremacola** [comment]
@elliotcondon the `render_block_preview` action only run at first init? I'm trying to update the block but I need to rerun `render_block_preview ` because some personal JS functions.

**Comment by elliotcondon** [comment]
Hi @andremacola 

Thanks for the feedback. Can you please share with me the full code that you are using to register your block and append this custom JS action? I'll set this up on my local site and diagnose the issue 👍

**Comment by andremacola** [comment]
@elliotcondon oh my god, I'm sorry. I was using window.acf.**doAction** instead window.acf.**addAction**  🤦‍♂️

**Comment by elliotcondon** [comment]
Ooh no! Well, let's hope you at least listened to some good music in that past hour 🎶🙌.
